### PR TITLE
Use 'unixsocket-http' for low-level calls.

### PR DESCRIPTION
**Pull request by xsc**
This introduces [unixsocket-http](https://github.com/into-docker/unixsocket-http) for low-level calls and removes the existing custom socket and factory implementations.

All existing tests pass. Are there any edge cases that need to be verified manually?

---

Generally, it feels to me that the note on low-level API access should be removed from the `README` and the whole `requests` namespace not be part of the public API. This would give more flexibility in making internal changes and would remove some of the mismatches between `fetch` and `unixsocket-http.core/request` (e.g. `:header` vs. `:headers`, `:query` vs `:query-params`).

Not in this PR of course, just looking for an opinion.

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -22,13 +22,9 @@
...
                  [clj-commons/clj-yaml "0.7.1"]
                  [metosin/jsonista "0.2.5"]
-                 [com.squareup.okhttp3/okhttp "4.5.0"]
-                 [com.github.jnr/jnr-unixsocket "0.30"]]
+                 [unixsocket-http "1.0.5"]]
   :plugins [[lein-ancient "0.6.15"]]
   :global-vars {*warn-on-reflection* true}
-  :aot [clj-docker-client.socket.TunnelingUnixSocket
-        clj-docker-client.socket.UnixDomainSocketFactory]
-  :resource-paths ["resources"]
```
Comment (lispyclouds) [inline project.clj:31]: We need the resources to pack the swagger YAMLs before deploying, I think we need need this as the default is something else?

**Comment by lispyclouds** [comment]
The primary reason the internal exposure was done to enable cases like the undocumented API access. This seems to be of use to people at https://nextjournal.com/ who use it for taking snapshots of the docker images via [checkpoints](https://docs.docker.com/engine/reference/commandline/checkpoint/) which seem to be not really exposed and experimental. Also in the previous versions of this lib people we trying to find ways of unsupported behaviour and I decided to just document a way.

May be can take the clojure apporach of marking these as internal in the doc and saying these are subject to change at anytime?

File: project.clj (hunk 1 above)
Comment (xsc) [reply project.clj:31]: Default should indeed be `resources`:
https://github.com/technomancy/leiningen/blob/master/leiningen-core/src/leiningen/core/project.clj#L221

Comment (lispyclouds) [reply project.clj:31]: Right, I wasnt sure if its resources or src/clj_docker_client/resources. good to know 😄

**Comment by lispyclouds** [comment]
In general this looks okay to me 😄

**Comment by lispyclouds** [comment]
Should I merge this? Maybe we can think of a way to address the low level issues soon.

**Comment by xsc** [comment]
It could also look like this:

- Make `connect` again part of the core API, let it return a unixsocket-http client. This might be a good thing anyways, since the underlying client now handles creating a new connection on each request (which I assume was the reason for this change in the first place).
- Document that it returns a unixsocket-http client.
- Adjust the README section to point that out and that you can use unixsocket-http directly to perform low-level calls.
- The only thing missing then would be the path interpolation.

I think it would be in the spirit of separation of concerns to let the lower-level stuff be handled by a lower-level library. However, documenting what the lower-level library is (and that `connect` returns such a client) creates coupling that might be a bad thing if we want to retain the flexibility to switch again at some point. (Or if we break backwards-compatibility in unixsocket-http for whatever reason.)

If something is changed in that regard, we'll have to have a transition period anyways, so there's enough time to think about it.

**Comment by xsc** [comment]
I have not dived into that part of the code yet, but I assume that there is some data structure that each `client` uses internally. Enabling experimental features could then be as simple as allowing users to "bring their own endpoints" (think writing their own YAML):

```clojure
(docker/client-from
  {:conn conn
   :endpoints [{:id :CreateCheckpoint, ... }] ;; request specification here
   :api-version "..."})
```

I mean, you could probably do that already by creating a file in `resources/api/my-own-version.yaml` in your own project, then setting `:api-version "my-own-version"`, right?

**Comment by lispyclouds** [comment]
> Make connect again part of the core API, let it return a unixsocket-http client. This might be a good thing anyways, since the underlying client now handles creating a new connection on each request (which I assume was the reason for this change in the first place).

Yes thats a correct assumption and I was having trouble with connection pooling + bidirectional sockets, but if with unixsocket-http thats fixed we may enable pooling again?

And saying that the lib is directly exposed if you'd wanna do internal things is okay and since the interpolation is available, users can use it.

> but I assume that there is some data structure that each client uses internally.

Yes its the parsed YAML pretty much which the lib walks through and figures out the param types and interpolation points. Nothing fancy there 😅 Also the interceptors might help here too.

> I mean, you could probably do that already by creating a file in resources/api/my-own-version.yaml in your own project, then setting :api-version "my-own-version", right?

Yes this can also be a valid solution. Not sure if the classpaths of the user and the lib jars are namespaced somehow?

**Comment by xsc** [comment]
> Yes this can also be a valid solution. Not sure if the classpaths of the user and the lib jars are namespaced somehow?

They are not, but they should be merged. However, I'd strongly advise to do namespacing by using at least `resources/clj-docker-client/` or even `resources/lispyclouds/clj-docker-client` to follow common advice that those things should be prefixed with `groupId/artifactId`.

**Comment by xsc** [comment]
> **Edit:** Or just allowing a `File`, `URI` or `InputStream` for `:api-version`.

You might not have seen the above since I added it afterwards. That would allow us to namespace our resources and make it super-simple for someone to plug in their own YAML.

**Comment by lispyclouds** [comment]
Nice! Fair points, I will look into this with a PR soon. Maybe time to cut a `1.0.0` release too? 🤔 There is [this](https://github.com/nextjournal/clj-docker-client/tree/async-invocations) I was really looking forward to but can merge whe it comes I guess.
